File: PrimePuzzlesPhone/src/main/java/net/geeksempire/primepuzzles/Utils/UI/SwipeGestureFilter.java
package net.geeksempire.primepuzzles.Utils.UI;

import android.content.Context;
import android.view.GestureDetector;
import android.view.GestureDetector.SimpleOnGestureListener;
import android.view.MotionEvent;
import android.view.View;

public class SwipeGestureFilter extends SimpleOnGestureListener {

    public final static int SWIPE_UP = 1;
    public final static int SWIPE_DOWN = 2;
    public final static int SWIPE_LEFT = 3;
    public final static int SWIPE_RIGHT = 4;

    public final static int MODE_TRANSPARENT = 0;
    public final static int MODE_SOLID = 1;
    public final static int MODE_DYNAMIC = 2;

    private final static int ACTION_FAKE = -666;

    private int swipe_Min_Distance = 10;
    private int swipe_Max_Distance = 1000;

    private int swipe_Min_Velocity = 10;

    private int mode = MODE_DYNAMIC;

    private boolean running = true;

    private boolean tapIndicator = false;

    private View view;
    private GestureDetector gestureDetector;
    private GestureListener gestureListener;

    public SwipeGestureFilter(View view, Context context, GestureListener gestureListener) {
        this.view = view;
        this.gestureDetector = new GestureDetector(context, this);
        this.gestureListener = gestureListener;
    }

    public void onTouchEvent(MotionEvent event) {

        if (!this.running) {
            return;
        }

        boolean result = this.gestureDetector.onTouchEvent(event);
        if (this.mode == MODE_SOLID) {
            event.setAction(MotionEvent.ACTION_CANCEL);
        }
        else if (this.mode == MODE_DYNAMIC) {
            if (event.getAction() == ACTION_FAKE) {
                event.setAction(MotionEvent.ACTION_UP);
            }
            else if (result) {
                event.setAction(MotionEvent.ACTION_CANCEL);
            }
            else if (this.tapIndicator) {
                event.setAction(MotionEvent.ACTION_DOWN);
                this.tapIndicator = false;
            }
        }
        //Else -> Just Do Nothing | MODE_TRANSPARENT
    }

    public int getMode() {
        return this.mode;
    }

    public void setMode(int mode) {
        this.mode = mode;
    }

    public void setEnabled(boolean status) {
        this.running = status;
    }

    public int getSwipeMaxDistance() {
        return this.swipe_Max_Distance;
    }

    public void setSwipeMaxDistance(int distance) {
        this.swipe_Max_Distance = distance;
    }

    public int getSwipeMinDistance() {
        return this.swipe_Min_Distance;
    }

    public void setSwipeMinDistance(int distance) {
        this.swipe_Min_Distance = distance;
    }

    public int getSwipeMinVelocity() {
        return this.swipe_Min_Velocity;
    }

    public void setSwipeMinVelocity(int distance) {
        this.swipe_Min_Velocity = distance;
    }

    @Override
    public boolean onFling(MotionEvent downMotionEvent, MotionEvent moveMotionEvent, float velocityX, float velocityY) {

        final float xDistance = Math.abs(downMotionEvent.getX() - moveMotionEvent.getX());
        final float yDistance = Math.abs(downMotionEvent.getY() - moveMotionEvent.getY());

        if (xDistance > this.swipe_Max_Distance || yDistance > this.swipe_Max_Distance) {
            return false;
        }

        velocityX = Math.abs(velocityX);
        velocityY = Math.abs(velocityY);

        boolean result = false;
        if ((velocityY >= this.swipe_Min_Velocity) && (yDistance > this.swipe_Min_Distance) && (xDistance < yDistance)) {//Vertical
            if (downMotionEvent.getY() > moveMotionEvent.getY()) {//Bottom -> Up
                this.gestureListener.onSwipe(SWIPE_UP);

                /*view.animate()
                    .x(downMotionEvent.getRawX() + xDistance)
                        .y(downMotionEvent.getRawY() + yDistance)
                        .setDuration(0)
                        .start();*/
            } else {//Up -> Bottom
                this.gestureListener.onSwipe(SWIPE_DOWN);
            }
            result = true;
        }

        if ((velocityX >= this.swipe_Min_Velocity) && (xDistance > this.swipe_Min_Distance) && (yDistance < xDistance)) {//Horizontal
            if (downMotionEvent.getX() > moveMotionEvent.getX()) {//Right -> Left
                this.gestureListener.onSwipe(SWIPE_LEFT);
            } else {//Left -> Right
                this.gestureListener.onSwipe(SWIPE_RIGHT);
            }
            result = true;
        }
        return result;
    }

    @Override
    public boolean onSingleTapConfirmed(MotionEvent motionEvent) {
        this.gestureListener.onSingleTapUp();

        return false;
    }

    public interface GestureListener {
        void onSwipe(int direction);

        void onSingleTapUp();
    }
}

File: PrimePuzzlesPhone/src/main/java/net/geeksempire/primepuzzles/Utils/UI/SwipeGestureFilter.kt
package net.geeksempire.primepuzzles.Utils.UI

import android.content.Context
import android.view.GestureDetector
import android.view.GestureDetector.SimpleOnGestureListener
import android.view.MotionEvent
import android.view.View
import androidx.dynamicanimation.animation.DynamicAnimation
import androidx.dynamicanimation.animation.FlingAnimation
import net.geeksempire.primepuzzles.Utils.FunctionsClass.FunctionsClassUI
import kotlin.math.abs

class SwipeGestureFilter(private val view: View, initContext: Context, private val gestureListener: GestureListener) : SimpleOnGestureListener() {

    private val context: Context = initContext

    lateinit var functionsClassUI: FunctionsClassUI

    var swipeMinDistance = 10
    var swipeMaxDistance = 1000

    var swipeMinVelocity = 10

    var mode = MODE_DYNAMIC

    private var running = true

    private var tapIndicator = false
    private val gestureDetector: GestureDetector

    init {
        this.gestureDetector = GestureDetector(initContext, this)

        functionsClassUI = FunctionsClassUI(initContext)
    }

    fun onTouchEvent(motionEvent: MotionEvent) {
        if (!this.running) {
            return
        }

        val result = this.gestureDetector.onTouchEvent(motionEvent)
        if (this.mode == MODE_SOLID) {
            motionEvent.action = MotionEvent.ACTION_CANCEL
        } else if (this.mode == MODE_DYNAMIC) {
            if (motionEvent.action == ACTION_FAKE) {
                motionEvent.action = MotionEvent.ACTION_UP
            } else if (result) {
                motionEvent.action = MotionEvent.ACTION_CANCEL
            } else if (this.tapIndicator) {
                motionEvent.action = MotionEvent.ACTION_DOWN
                this.tapIndicator = false
            }
        }
        //Else -> Just Do Nothing | MODE_TRANSPARENT
    }

    fun setEnabled(status: Boolean) {
        this.running = status
    }

    override fun onFling(
        downMotionEvent: MotionEvent,
        moveMotionEvent: MotionEvent,
        initVelocityX: Float,
        initVelocityY: Float): Boolean {

        val velocityX: Float = initVelocityX
        val velocityY: Float = initVelocityY

        val xDistance = abs(downMotionEvent.x - moveMotionEvent.x)
        val yDistance = abs(downMotionEvent.y - moveMotionEvent.y)

        if (xDistance > this.swipeMaxDistance || yDistance > this.swipeMaxDistance) {
            return false
        }

        val flingAnimationX: FlingAnimation by lazy(LazyThreadSafetyMode.NONE) {
            FlingAnimation(view, DynamicAnimation.X)
                .setFriction(1.3f)
                .setMinValue(0f)
                .setMaxValue(context.resources.displayMetrics.widthPixels.toFloat())
        }
        val flingAnimationY: FlingAnimation by lazy(LazyThreadSafetyMode.NONE) {
            FlingAnimation(view, DynamicAnimation.Y)
                .setFriction(1.3f)
                .setMinValue(0f)
                .setMaxValue(context.resources.displayMetrics.heightPixels.toFloat())
        }

        var result = false
        if (abs(velocityY) >= this.swipeMinVelocity && yDistance > this.swipeMinDistance && xDistance < yDistance) {//Vertical
            if (downMotionEvent.y > moveMotionEvent.y) {//Bottom -> Up
                this.gestureListener.onSwipe(SWIPE_UP)



            } else {//Up -> Bottom
                this.gestureListener.onSwipe(SWIPE_DOWN)



            }

            flingAnimationY.setStartValue(view.y)
            flingAnimationY.setStartVelocity(velocityY)
            flingAnimationY.start()

            result = true
        }

        if (abs(velocityX) >= this.swipeMinVelocity && xDistance > this.swipeMinDistance && yDistance < xDistance) {//Horizontal
            if (downMotionEvent.x > moveMotionEvent.x) {//Right -> Left
                this.gestureListener.onSwipe(SWIPE_LEFT)



            } else {//Left -> Right
                this.gestureListener.onSwipe(SWIPE_RIGHT)



            }

            flingAnimationX.setStartValue(view.x)
            flingAnimationX.setStartVelocity(velocityX)
            flingAnimationX.start()

            result = true
        }
        return result
    }

    override fun onSingleTapConfirmed(motionEvent: MotionEvent): Boolean {
        this.gestureListener.onSingleTapUp()

        return false
    }

    interface GestureListener {
        fun onSwipe(direction: Int)

        fun onSingleTapUp()
    }

    companion object {
        const val SWIPE_UP = 1
        const val SWIPE_DOWN = 2
        const val SWIPE_LEFT = 3
        const val SWIPE_RIGHT = 4

        const val MODE_TRANSPARENT = 0
        const val MODE_SOLID = 1
        const val MODE_DYNAMIC = 2

        private val ACTION_FAKE = -666
    }
}
